### DISCUSSION TITLE
Is it possible to use React-based components within a NiceGUI app?

### DISCUSSION THREAD

**Discussion by EmilioGarCan** [zauberzeug/nicegui#3720] [Q&A]
### Question

I'm looking to integrate a calendar heatmap into my app. I've tried the echarts calendar heatmap but found it to be too inflexible for my use case.

I found [cal-heatmap](https://cal-heatmap.com/docs/getting-started/installation), it looks cool and the installation seemed simple, but then I saw that it uses React (the render function). I know little to nothing about React, but I would still like to implement this component into my NiceGUI app. 

Is this possible? How would I begin to approach this problem? Appreciate any help.

**Comment by falkoschindler** [comment]
@EmilioGarCan As far as I can tell, [it works with React](https://cal-heatmap.com/docs/faq#does-it-work-with-react-), but it doesn't require it. So you might be able to integrate it as a [custom component](https://nicegui.io/documentation/section_configuration_deployment#custom_vue_components).

**Comment by puntorigen** [comment]
Yes — this is exactly what the [`nicegui-react`](https://github.com/puntorigen/nicegui-react) package is for. It builds your React project with Vite, serves the bundle, and mounts it inside a native NiceGUI component, so props flow Python → React and events flow React → Python over NiceGUI's websocket (reactive in production too, not just dev).

Minimal example:

```python
from nicegui import ui
from nicegui_react import React

@ui.page('/')
def index():
    chart = React('./my_react_project', main_component='App').props(title='Hello from Python!')
    chart.on('onClick', lambda data: ui.notify(f'React said: {data}'))

ui.run()
```

Any npm dependency your component imports is bundled automatically (recharts, etc.), and you can place multiple different components / instances on a single page.

`pip install nicegui-react`

(For cal-heatmap specifically: as noted above it can run without React, so a plain custom Vue component may be lighter — but if you'd rather use the React render path or any other React lib, this handles it.)
